Assertthat Plugin Extension

GaneshSPatil opened the pull request:
Hello @goloroden,

This is an initial draft with for the plugin API extension.

The extension endpoint is documented [here](https://github.com/GaneshSPatil/assertthat/wiki/Asserthat-Plugin-Extension).

I have written an example [assertthat-type](https://github.com/GaneshSPatil/assertthat-type) plugin to demonstrate the use case of the plugin extension.

I would love to hear some feedback on the same.

#### Pending Tasks:
- Some more integration testing around the extension itself.
- Haven't done exception handling. 
    - Example: Accessing extension which hasn't added.
    - Example: Adding duplicate extension/methods by the same plugin.
    - and many more...
- Not sure how to generate `dist` code.



#### Some open questions:
  - The extension exposes `assert.fail` method. (uhh!! any better way to get the handle of assertion failure)?
  - The `addMethod` expects to return a function. The returned function is expected to perform assertions. (As we know, its due to the way `assertthat` has been designed.) Is there any other way to simplify extension endpoint method?

  - The API extension methods which are visible for `assertthat` users too. (can we rename the methods to start with an underscore? Example: `assert._forExtension`).

  - Should we restrict override native methods?

goloroden commented:
Hi @GaneshSPatil 😊 

Thanks for your PR. Currently I am pretty ill, hence I don't have the mind to think about this, but I'll get back to you as soon as possible!

GaneshSPatil commented:
Sure @goloroden,

Take care!

goloroden commented:
Hi @GaneshSPatil 😊 

I'm feeling somewhat better, so at least I can give some feedback to you now!

First of all, thanks a lot for your hard work, and all of your ideas that you have put into this extension! I like the idea of a plugin concept a lot, and I think that your current setup is a good place to start.

There are a few things that caught my eye:

The way to connect a plugin to assertthat feels strange to me. Right now assertthat is an option for the plugin, this is IMHO counterintuitive. What do you think about introducing a `use` function on `assert` that you can hand over a plugin? Something such as:

```javascript
assert.use(myPlugin);
```

I think this is a well-known pattern in the Node.js world.

Internally, all the `use` function would ne to do was to call an `initialize` function on the plugin, and hand over itself (the `assert` object), so that you can call the plugin methods such as `addMethod`, `addExtension`, …

What do you think?

GaneshSPatil commented:
@goloroden -- Glad to know that you are doing good!

Agree with you that the endpoint interaction change. It should be changed to `assert.use`.

In my previous attempt to add plugin extension I used the same approach. [here](https://github.com/GaneshSPatil/assertthat-type-old#plug-in). I quiet not remember the reason behind moving away from `assert.use` :(

GaneshSPatil commented:
Hello @goloroden ,

I've fixed the extension to have an add method through which plugin could be added.
It also fixed the problem of plugin having the entire control on assert. Now plugin gets access to three extension methods: `forExtension`, `addExtension` and `fail`. Which ensures that  a plugin can't damage the core functionality of `assertthat`.
